File: c++/incomplete.cpp
// https://programmers.co.kr/learn/courses/30/lessons/42576?language=cpp

#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

string solution(vector<string> participant, vector<string> completion)
{
    unordered_multiset<string> ms;
    for (string part : participant)
    {
        ms.insert(part);
    }

    for (string comp : completion)
    {
        auto itr = ms.find(comp);
        if (itr != ms.end())
        {
            ms.erase(itr);
        }
    }

    return *ms.begin();
}

// string solution(vector<string> participant, vector<string> completion)
// {
//     unordered_multiset<string> partSet, compSet;

//     const int n = participant.size();
//     for (int i = 0; i < n - 1; i++)
//     {
//         partSet.insert(participant[i]);
//         compSet.insert(completion[i]);
//     }
//     partSet.insert(participant[n - 1]);

//     string answer = "";
//     for (string part : participant)
//     {
//         if (partSet.count(part) != compSet.count(part))
//         {
//             answer = part;
//             break;
//         }
//     }

//     return answer;
// }

File: python/incomplete.py
# https://programmers.co.kr/learn/courses/30/lessons/42576?language=python3

def solution(participant, completion):
    participant.sort()
    completion.sort()

    for i in range(len(completion)):
        if participant[i] != completion[i]:
            return participant[i]
    return participant[-1]


# Similar, but using zip
# def solution(participant, completion):
#     participant.sort()
#     completion.sort()
#     for p, c in zip(participant, completion):
#         if p != c:
#             return p
#     return participant[-1]

# Using collections
# import collections

# def solution(participant, completion):
#     answer = collections.Counter(participant) - collections.Counter(completion)
#     return list(answer.keys())[0]


# Using hash
# def solution(participant, completion):
#     answer = ''
#     temp = 0
#     dic = {}
#     for part in participant:
#         dic[hash(part)] = part
#         temp += int(hash(part))
#     for com in completion:
#         temp -= hash(com)
#     answer = dic[temp]

#     return answer
